engines/wage/design.cpp

```diff
@@ -0,0 +1,539 @@
+		}
+		xcoords.push_back(x1);
+		ycoords.push_back(y1);
+		x1 = x2;
+		y1 = y2;
+	}
+	xcoords.push_back(x1);
+	ycoords.push_back(y1);
+
+	int npoints = xcoords.size();
+	int *xpoints = (int *)calloc(npoints, sizeof(int));
+	int *ypoints = (int *)calloc(npoints, sizeof(int));
+	for (int i = 0; i < npoints; i++) {
+		xpoints[i] = xcoords[i];
+		ypoints[i] = ycoords[i];
+	}
+
+	plotData pd(surface, &patterns, fillType, 1);
+
+	if (fillType <= patterns.size()) {
+		Graphics::drawPolygonScan(xpoints, ypoints, npoints, bbox, kColorBlack, drawPixel, &pd);
+	}
+
+	pd.fillType = borderFillType;
+	pd.thickness = borderThickness;
+	if (borderThickness > 0 && borderFillType <= patterns.size()) {
+		for (int i = 1; i < npoints; i++)
+			Graphics::drawLine(xpoints[i-1], ypoints[i-1], xpoints[i], ypoints[i], kColorBlack, drawPixel, &pd);
+	}
+
+	free(xpoints);
+	free(ypoints);
+}
+
+void Design::drawOval(Graphics::Surface *surface, Common::ReadStream &in,
+			Patterns &patterns, byte fillType, byte borderThickness, byte borderFillType) {
+	int16 y1 = in.readSint16BE();
+	int16 x1 = in.readSint16BE();
+	int16 y2 = in.readSint16BE();
+	int16 x2 = in.readSint16BE();
+	plotData pd(surface, &patterns, fillType, 1);
+
+	if (fillType <= patterns.size())
+		Graphics::drawEllipse(x1, y1, x2-1, y2-1, kColorBlack, true, drawPixel, &pd);
+
+	pd.fillType = borderFillType;
+	pd.thickness = borderThickness;
+
+	if (borderThickness > 0 && borderFillType <= patterns.size())
+		Graphics::drawEllipse(x1, y1, x2-1, y2-1, kColorBlack, false, drawPixel, &pd);
+}
+
+void Design::drawBitmap(Graphics::Surface *surface, Common::ReadStream &in) {
+	int numBytes = in.readSint16BE();
+	int y1 = in.readSint16BE();
+	int x1 = in.readSint16BE();
+	int y2 = in.readSint16BE();
+	int x2 = in.readSint16BE();
+	int w = x2 - x1;
+	int h = y2 - y1;
+	Graphics::Surface tmp;
+
+	tmp.create(w, h, Graphics::PixelFormat::createFormatCLUT8());
+
+	numBytes -= 10;
+
+	int x = 0, y = 0;
+	while (numBytes > 0 && y < h) {
+		int n = in.readSByte();
+		int count;
+		int b;
+		int state = 0;
+
+		numBytes--;
+
+		if ((n >= 0) && (n <= 127)) { // If n is between 0 and 127 inclusive, copy the next n+1 bytes literally.
+			count = n + 1;
+			state = 1;
+		} else if ((n >= -127) && (n <= -1)) { // Else if n is between -127 and -1 inclusive, copy the next byte -n+1 times.
+			b = in.readByte();
+			numBytes--;
+			count = -n + 1;
+			state = 2;
+		} else { // Else if n is -128, noop.
+			count = 0;
+		}
+
+		for (int i = 0; i < count && y < h; i++) {
+			byte color;
+			if (state == 1) {
+				color = in.readByte();
+				numBytes--;
+			} else if (state == 2)
+				color = b;
+
+			for (int c = 0; c < 8; c++) {
+				if (x1 + x >= 0 && x1 + x < surface->w && y1 + y >= 0 && y1 + y < surface->h)
+					*((byte *)tmp.getBasePtr(x, y)) = (color & (1 << (7 - c % 8))) ? kColorBlack : kColorWhite;
+				x++;
+				if (x == w) {
+					y++;
+					x = 0;
+					break;
+				}
+			}
+		}
+	}
+
+	while (numBytes--)
+		in.readByte();
```

Comment:
It looks like `drawBitmap` is always used on a `SeekableReadStream`, this would allow to make `in` of such type and thus allow us to use `skip`/`seek` here.